Assets/MixedRealityToolkit.Services/InputAnimation/InputAnimation.cs

```diff
@@ -0,0 +1,710 @@
+            if (iX > 0 && iY > 0 && iZ > 0)
+            {
+                Vector3 v0 = new Vector3(curveX.keys[iX - 1].value, curveY.keys[iY - 1].value, curveZ.keys[iZ - 1].value);
+                Vector3 v1 = new Vector3(curveX.keys[iX].value, curveY.keys[iY].value, curveZ.keys[iZ].value);
+
+                // Merge the preceding two intervals if difference is small enough
+                if ((v1 - v0).sqrMagnitude <= sqrThreshold && (position - v1).sqrMagnitude <= sqrThreshold)
+                {
+                    curveX.RemoveKey(iX);
+                    curveY.RemoveKey(iY);
+                    curveZ.RemoveKey(iZ);
+                }
+            }
+
+            AddFloatKey(curveX, time, position.x);
+            AddFloatKey(curveY, time, position.y);
+            AddFloatKey(curveZ, time, position.z);
+        }
+
+        private static void AddRotationKeyFiltered(AnimationCurve curveX, AnimationCurve curveY, AnimationCurve curveZ, AnimationCurve curveW, float time, Quaternion rotation, float threshold)
+        {
+            float sqrThreshold = threshold * threshold;
+
+            int iX = FindKeyframeInterval(curveX, time);
+            int iY = FindKeyframeInterval(curveY, time);
+            int iZ = FindKeyframeInterval(curveZ, time);
+            int iW = FindKeyframeInterval(curveW, time);
+            if (iX > 0 && iY > 0 && iZ > 0 && iW > 0)
+            {
+                Quaternion q0 = new Quaternion(curveX.keys[iX - 1].value, curveY.keys[iY - 1].value, curveZ.keys[iZ - 1].value, curveW.keys[iW - 1].value);
+                Quaternion q1 = new Quaternion(curveX.keys[iX].value, curveY.keys[iY].value, curveZ.keys[iZ].value, curveW.keys[iW].value);
+
+                // Merge the preceding two intervals if difference is small enough
+                (q0 * Quaternion.Inverse(q1)).ToAngleAxis(out float angle0, out Vector3 axis0);
+                (rotation * Quaternion.Inverse(q0)).ToAngleAxis(out float angle1, out Vector3 axis1);
+                if (angle0 <= sqrThreshold && angle1 <= sqrThreshold)
+                {
+                    curveX.RemoveKey(iX);
+                    curveY.RemoveKey(iY);
+                    curveZ.RemoveKey(iZ);
+                    curveW.RemoveKey(iW);
+                }
+            }
+
+            // TODO make use of Bezier interpolation to allow more aggressive compression
+            // and use tangents and weights to accurately merge adjacent splines.
+            AddFloatKey(curveX, time, rotation.x);
+            AddFloatKey(curveY, time, rotation.y);
+            AddFloatKey(curveZ, time, rotation.z);
+            AddFloatKey(curveW, time, rotation.w);
+        }
+
+        /// Add a float value to an animation curve
+        private static int AddFloatKey(AnimationCurve curve, float time, float value)
+        {
+            // return curve.AddKey(time, value);
+
+            // Use linear interpolation to avoid overshooting from bezier tangents
```

Comment:
I'm new to this but can you show me where the lerping is occuring here? I would have expected to see some sort of lerp call (or some non-zero keyframe params or something that might be like a lerp)

Reply:
Or is it just that, the lerp occurs over time to the specified value

Reply:
The Unity Keyframe ctor takes values for tangents and weights for Bezier interpolation. I'm only creating linear interpolation keyframes though, by setting tangents and weights to zero.

I thought about utilizing additional interpolation features of Bezier splines to accurately record movement with fewer keyframes, but i don't have time to figure out the math for that and linear interpolation does the job for us. For reducing file size it's more important to have a sane binary format (e.g. glTF).

I've updated comments to explain a bit better.

If you feel like diving into Bezier curve math i can recommend this site: [https://pomax.github.io/bezierinfo/](https://pomax.github.io/bezierinfo/)